js/private/js_library_helpers.bzl

```diff
@@ -0,0 +1,244 @@
+                if ctx.label.package != file.owner.package:
+                    msg = """
+
+Expected to find source file {file_basename} in {this_package}, but instead it is in {file_package}.
+
+All source files in rules_js rules must be in the same package as the target.
+
+See https://github.com/aspect-build/rules_js/tree/67c561e08fa8baaf6fbfe00094fa0d757159e1ef#running-nodejs-programs
+for more context on why this is required.
+
+Either move {file_basename} to {this_package}, or add a 'js_library'
+target in {file_basename}'s package and add that target to the deps of {this_target}:
+
+    buildozer 'new_load @aspect_rules_js//js:defs.bzl js_library' {file_package}:__pkg__
+    buildozer 'new js_library {new_target_name}' {file_package}:__pkg__
+    buildozer 'add srcs {file_basename}' {file_package}:{new_target_name}
+    buildozer 'add visibility {this_package}:__pkg__' {file_package}:{new_target_name}
+    buildozer 'remove srcs {file_package}:{file_basename}' {this_target}
+    buildozer 'add srcs {file_package}:{new_target_name}' {this_target}
+
+""".format(
+                        file_basename = file.basename,
+                        file_package = "%s//%s" % (file.owner.workspace_name, file.owner.package),
+                        new_target_name = file.basename.replace(".", "_"),
+                        this_package = "%s//%s" % (ctx.label.workspace_name, ctx.label.package),
+                        this_target = ctx.label,
+                    )
+                    fail(msg)
+                file = copy_file_to_bin_action(ctx, file, is_windows = is_windows)
+
+            if file.is_directory:
+                # assume a directory contains declarations since we can't know that it doesn't
+                declarations.append(file)
+                sources.append(file)
+            elif (
+                file.path.endswith(".d.ts") or
+                file.path.endswith(".d.ts.map") or
+                file.path.endswith(".d.mts") or
+                file.path.endswith(".d.mts.map") or
+                file.path.endswith(".d.cts") or
+                file.path.endswith(".d.cts.map")
+            ):
+                declarations.append(file)
+            elif file.path.endswith("/package.json"):
+                # package.json may be required to resolve declarations with the "typings" key
+                declarations.append(file)
+                sources.append(file)
+            else:
+                sources.append(file)
+
+    # sources
+    sources.extend([
+        item
+        for target in targets
+        if JsInfo in target
+        for item in target[JsInfo].sources
+    ])
+    sources = [file for file in sources if file]
+
+    # declarations
+    declarations.extend([
+        item
+        for target in targets
+        if JsInfo in target
+        for item in target[JsInfo].declarations
+    ])
+    declarations = [file for file in declarations if file]
+
+    return (sources, declarations)
```

Comment:
do you have a reason for returning tuples rather than structs? the latter seems more common and idiomatic

Reply:
good point, I'll switch to structs